### ISSUE TITLE
Expected behaviour when a tiddler open in edit mode is modifed from outside the draft?

### ISSUE THREAD

**Issue by erwanm** [Jermolene/TiddlyWiki5#2108]
# Description 

1. the user opens a tiddler A for edition
2. the user does something else with some other tiddlers (maybe forgets about the open draft). For instance they click a button which changes the value of some field in tiddler A
3. Later they come back to the draft of tiddler A and save the changes

Problem: the value of the field modifed at step 2 is overwritten.

## Example

on tiddlywiki.com create a new tiddler with this content:

```<$button>
<$action-setfield $tiddler="GettingStarted" tags="[[new tag!]]"/>
<$action-navigate $to="GettingStarted"/>
Modify tag of GettingStarted
</$button>
```

then open GettingStarted for edition, and then click the button from the new tiddler: at this stage the tags in the non-draft version of GettingStarted contains "new tag!". Now if you close the draft version by clicking the save button, the tags go back to the previous version. (additionally GettingStarted is open twice in the story river, which is strange).

# Possible ways to fix

(simple suggestions that come to my mind)

   * a warning message when clicking "save" if some modifications have been done since the draft was opened 
   * a warning any time a non-draft tiddler is modified while the draft version is open
   * prevent any more editing on the draft until the user chooses which version they want to keep: either they reload the modified non-draft version and lose any change done in edit mode, or they keep the draft changes and lose the other modifications. 
      * I know at least one text editor which behave like this: if the open file is not in sync anymore with the file on disk, the user has to select which version they want before continuing.
   * there is also the possibility to synchronize silently the draft version, but this might cause more problems.

Remark: I assume that this case would mostly happen by mistake, when the user forgets they have opened a tiddler. The case of a user saving their wiki with a draft open is worth mentioning, since in this case they cannot see it in the open tiddlers tab at the next session.

# Related problem

The original problem I faced was when I was trying to have a control which works in both view and draft mode: my first idea was to alway apply it to the non-draft tiddler, but then it wouldn't work in edit mode. See details here: https://groups.google.com/d/msg/tiddlywiki/crkF76agkTI/pvDiObKNCAAJ

**Comment by felixhayashi** [comment]
I agree with @erwanm that this is an important topic many people are not aware of.

> a warning message when clicking "save" if some modifications have been done since the draft was opened 

I think it is not possible to figure out which fields have changed because you don't know if it was changed in the original tiddler or in the draft.

But: TW could still compare the `creation` date of the draft tiddler with the `modified` date of the original tiddler. If the original tiddler has a modified timestamp that is later than the created timestamp of the draft, TW should imo display a warning.

> a warning any time a non-draft tiddler is modified while the draft version is open

I'd rather not fire dialogs in this case; But what would be great is a silent but visible banner at the top of the tiddler that says something like:

![editwarning](https://cloud.githubusercontent.com/assets/4307137/11651510/9439dcda-9d91-11e5-915b-dfe3b1098787.png)

The banner is only revealed if `original.modified > draft.created`.

-Felix

**Comment by tobibeer** [comment]
@felixhayashi Great idea!

**Comment by Jermolene** [comment]
Hi @erwanm this is actually the same problem as handling multiple users simultaneously editing the same wiki. My wiki experience suggests that the optimum user experience is pretty much what you've agreed here:

1. Show a visible warning if another user (or the system) modifies a tiddler while it is being edited
2. Allow the user to review the change and to opt to replace the current draft with the newly modified values
3. If the user chooses to proceed with saving their changes, they should overwrite the newly changed value
4. The user should be able to use revision history to retrospectively repair clashes

The easiest way to keep track of whether the target tiddler has been modified is to use the existing change count mechanism that is built into the store. We'd add a `draft.original` field that contains the changecount of the target tiddler at the time the draft was created.

The trouble with that approach is that changecounts are not persistent across sessions (they were first implemented to support synchronisation).

One alternative is for the `draft.original` field to contain a string representation of the original tiddler values (eg, a JSON string, or a hash of a JSON string).
